docs/exceptions.md

```diff
@@ -192,13 +239,21 @@ Do:
 * create a custom exception type
 * use `throwM` (`MonadThrow`)
 
+*Comment:* the last bullet in "Do not" is a silly rule and should be struck.
+If the `Left` case is a normal outcome, then leave it be! Making it into an
+exception and throwing it is unambiguously worse. The type is less descriptive,
+and the normal case is mixed in with every truly exceptional case. It's simply
+a lot harder for a caller to use the function properly.
```

Comment:
It is easier to define and use this function properly:

* the caller does not have to account for two ways of exception throwing
* the implementor does not have the freedom to make a mistake and use `throwM` or `throwIO` instead of `throwError`

The assumption you seem to make is that `IO` exceptions aren't used for regular errors, this implies we never want to catch and handle them. That's simply not how other libraries are structured, even `base` throws exceptions for normal outcomes (see `readFile`), we have to live with that and the guidelines are written accordingly.

If we used some cunning technique to represent all regular errors differently, I wouldn't mind (I already cited `unexceptional-io`), but at the moment we don't.

Reply:
> the caller does not have to account for two ways of exception throwing

I'm not asking for 2 ways of exception throwing. I'm asking that we don't throw "exceptions" that aren't actually exceptional. Maximize purity. Not every `Left` or `Nothing` should be treated like an exception or a failure.

> the implementor does not have the freedom to make a mistake and use throwM or throwIO instead of throwError

And the implementor wouldn't do that anyway.

> The assumption you seem to make is that IO exceptions aren't used for regular errors, this implies we never want to catch and handle them. 

I don't understand what you mean. Isn't a "regular error" the same as an "exception"? Please, "regular error" has to go. My assumption is that `IO` exceptions are (sometimes) used to express exceptions. My advice is that `IO` exceptions should be used sparingly.

Reply:
> I don't understand what you mean. Isn't a "regular error" the same as an "exception"? Please, "regular error" has to go. 

The document gives a decision procedure for whether an error should be classified as a regular error (not necessarily an exception) or a programmer mistake (bug).

>  I'm asking that we don't throw "exceptions" that aren't actually exceptional. Maximize purity. 

I assume something can be classified as exceptional when we never want to `catch` it to affect the logic of our app, only to release some resources and rethrow (`bracket`). Based on this understanding, I have only one objection:

* existing libraries don't follow this principle, at all

If you're using a different definition of "exceptional", then it's not clear to me what you mean.